metsim/disaggregate.py

```diff
@@ -7,107 +7,179 @@
-    df_disagg['longwave'] = longwave(df_daily['lwrad'])
-    df_disagg['wind'] = wind(df_daily['wind'])
+                                        df_daily['dayl'],
+                                        df_daily['day_of_year'],
+                                        solar_geom['tiny_rad_fract'],
+                                        params))
+    
+    t_Tmin, t_Tmax = set_min_max_hour(df_disagg['shortwave'], n_days, ts)
+    df_disagg['temp'] = temp(df_daily, df_disagg, t_Tmin, t_Tmax, ts)
+    
+    df_disagg['vapor_pressure'] = vapor_pressure(df_daily['vapor_pressure'],
+                                                 df_disagg['temp'],
+                                                 t_Tmin, n_disagg, ts)
+    
+    df_disagg['rel_humid'] = relative_humidity(df_disagg['vapor_pressure'],
+                                               df_disagg['temp'])
+   
+    df_disagg['longwave'], df_disagg['tskc'] = longwave(df_disagg['temp'],
+                                                        df_disagg['vapor_pressure'],
+                                                        df_daily['tskc'])
+
+    df_disagg['precip'] = precip(df_daily['precip'], ts)
+    df_disagg['wind'] = wind(df_daily['wind'], ts)
+    df_disagg['pet'] = df_daily['pet'].resample(str(int(ts))+'T').fillna(method='ffill')
+
     return df_disagg.fillna(method='ffill')
 
 
-def set_min_max_hour(disagg_rad, n_days):
+def set_min_max_hour(disagg_rad, n_days, ts):
     """
     TODO
     """
-    ts = float(params['time_step'])
     rad_mask = 1*(disagg_rad > 0)
     diff_mask = np.diff(rad_mask)
     rise_times = np.where(diff_mask>0)[0] * ts 
     set_times = np.where(diff_mask<0)[0] * ts
     t_Tmax = 0.67 * (set_times - rise_times) + rise_times
-    t_Tmin = rise_times - float(params['time_step'])
+    t_Tmin = rise_times 
     return t_Tmin, t_Tmax
 
 
-def temp(df_daily, df_disagg):
+def temp(df_daily, df_disagg, t_Tmin, t_Tmax, ts):
     """
     TODO
     """
     # Calculate times of min/max temps
-    n_days = len(df_daily['day_of_year'])
-    t_Tmin, t_Tmax = set_min_max_hour(df_disagg['shortwave'], n_days)
     time = np.array(list(next(it) for it in itertools.cycle(
-                [iter(t_Tmin), iter(t_Tmax)])))
+                [iter(t_Tmin), iter(t_Tmax)])))  
     temp = np.array(list(next(it) for it in itertools.cycle(
                 [iter(df_daily['t_min']), iter(df_daily['t_max'])])))
+    # Account for end points
+    ts_ends = cnst.MIN_PER_HOUR * cnst.HOURS_PER_DAY
+    time = np.append(np.insert(time, 0, time[0:2]-ts_ends), time[-2:]+ts_ends)
+    temp = np.append(np.insert(temp, 0, temp[0:2]), temp[-2:])
     
+    # Interpolate the values
     try:
         interp = scipy.interpolate.PchipInterpolator(time, temp, extrapolate=True)
-        temps = interp(float(params['time_step']) * np.arange(0, len(df_disagg.index)))
+        temps = interp(ts * np.arange(0, len(df_disagg.index)))
     except ValueError:
```

Comment:
Good question... this may not even be the long term desired behavior.  The call can fail when the call to `PchipInterpolator` is given a timeseries that is not strictly increasing, or the `time` and `temp`  arguments have mismatched lengths.